Runner uninstall enhancements

paladin-devops opened the pull request:
This PR makes a few updates to `runner uninstall`, some general, and some specific to ECS. This addresses #3677. 

Regarding the general use of `runner uninstall`, if the install operation fails, a warning message is now output to let the user know that they can _attempt_ to uninstall the runner (which failed to be installed) with `runner uninstall`:

```terminal
$ waypoint runner install -platform=ecs -server-addr=<redacted>:9701 -server-tls-skip-verify -ecs-cluster=paladin-devops-cluster -ecs-region=ca-central-1
✓ Finished connecting to: <redacted>:9701
❌ Installing runner...
✓ Networking setup
✓ EFS ready
✓ Found existing IAM role to use: waypoint-runner-execution-role
✓ Found existing IAM role to use: waypoint-runner
✓ Using existing log group
❌ Using existing security group: waypoint-server-security-group
! Error installing runner: ClusterNotFoundException: Cluster not found.
Please run the following to clean up the resources from the unsuccessful runner installation,
specifying additional platform flags as needed:

waypoint runner uninstall -platform=ecs -id=01GE2MABHDJ7JPXGE5JDPTH86R <additional_platform_flags>
```

The other change being made for the general usage of `runner uninstall` is that we will check if a runner is registered to the server before attempting to forget it.

For ECS specifically, this PR adds a tag to the EFS volume which is created for the runner, and the tag contains the ID of the runner. When we uninstall the runner now, the EFS volume will also be deleted.

The log group, security group, IAM task role, and other resources are still **NOT** deleted, because they may be shared by other runners.

Tested scenarios:
- [x] runner install with existing server in ECS
- [x] runner uninstall with existing server in ECS
- [x] server install w/static runner to ECS
- [x] server uninstall w/static runner from ECS
- [ ] server upgrade w/static runner, from v0.10.1 in ECS

paladin-devops committed:
Output warning message if runner install fails with a command to uninstall the runner.

paladin-devops committed:
Update runner uninstall to not fail if the runner isn't registered to the server.

This will permit runners whose installations failed with connecting to the server to still be uninstalled from the platform, as they won't need to be forgotten by the server.

paladin-devops committed:
Update ECS runner uninstall.

Delete the EFS volume created during the runner install. Update the security group for the runner which is installed to not include ports which were required only for the server.

paladin-devops committed:
Specify different tags for EFS for runner install.

paladin-devops committed:
Update EFS deletion during ECS runner uninstall to allow AWS 5 minutes to delete the file system's mount targets.

paladin-devops committed:
Use terminal UI output for runner uninstall failure message.

paladin-devops committed:
Fix ECS runner uninstall error message format.

paladin-devops committed:
Changelog.

briancain commented on internal/cli/runner_install.go:
```diff
@@ -268,6 +268,7 @@ func (c *RunnerInstallCommand) Run(args []string) int {
 		c.ui.Output("Error installing runner: %s", clierrors.Humanize(err),
 			terminal.WithErrorStyle(),
 		)
+		c.ui.Output(runnerInstallFailed, c.platform[0], id, terminal.WithWarningStyle())
```
Pretty sure it should be this? Otherwise I think it's the same platform?

```suggestion
		c.ui.Output(runnerInstallFailed, platform, id, terminal.WithWarningStyle())
```

Or hrm, why is `c.platform` a slice string? Does runner install allow for installing to multiple platforms?

paladin-devops replied:
@briancain it's a slice because this var is an `EnumVar` 😄 

```go
                f.EnumVar(&flag.EnumVar{
			Name:   "platform",
			Usage:  "Platform to install the Waypoint runner into. If unset, uses the platform of the local context.",
			Values: sortedPlatformNames,
			Target: &c.platform,
		})
```

briancain replied:
Strange! Yeah I see EnumVar's target is a slice, but it's only ever valid for 1 flag haha.

briancain commented on internal/cli/runner_uninstall.go:
```diff
@@ -149,17 +147,30 @@ func (c *RunnerUninstallCommand) Run(args []string) int {
+			})
+			if err != nil {
+				s.Update("Couldn't forget runner: %s", clierrors.Humanize(err))
+				s.Status(terminal.StatusWarn)
```
This should probably be a ui.Output. Otherwise the next iteration, for either case, it'll get wiped due to how StepGroups work and the err will be lost.

catsby replied:
@briancain wouldn't that next iteration only happen if 2 runners have the same ID?

briancain replied:
Ah true, should we `break` out of this loop them for the `if runner.Id == c.id` case?

paladin-devops replied:
+1 to return 1, and I added a `break` at the conclusion of the `if` statement, since we don't need to continue looping if we found and successfully forgot the runner.

briancain commented on internal/cli/runner_uninstall.go:
```diff
@@ -149,17 +147,30 @@ func (c *RunnerUninstallCommand) Run(args []string) int {
+			s.Done()
+		}
 	}
-	s.Done()
```
We might still need this done, especially in the case where the length of `runners.Runners` is 0?

catsby commented on internal/serverinstall/ecs.go:
```diff
@@ -158,7 +158,14 @@ func (i *ECSInstaller) Install(
+			httpPort, _ := strconv.Atoi(serverconfig.DefaultHTTPPort)
+			ports := []*int64{
+				aws.Int64(int64(grpcPort)),
+				aws.Int64(int64(httpPort)), // TODO: Not needed for runner install
```
What is `TODO` about this? Just curious. The Runner install doesn't doesn't seem to include it or the `grpcPort`

paladin-devops replied:
The ports set here ultimately result in a security group being created that is shared by both the server and the runner. The server needs the HTTP port but the runner has no use for it. The TODO is getting at that we might want to split them up somehow, just to tighten the SG a bit.

catsby replied:
👍

catsby reviewed (changes requested):
Questions and suggestions specifically with efs deleting

paladin-devops committed:
fix: fail on runner uninstall if runner forget fails.

Also exit loop early if we successfully forget the runner to save a bit of time.

paladin-devops committed:
fix: End step in SG properly when forgetting runners during runner install.

paladin-devops committed:
fix: Account for pagination of results of file system query.

The EFS API for AWS returns paginated results if there are more than the specified # of results (100). This commit updates the query for existing file systems to account for this pagination by looping through the results, and checking if the marker in the respone of the query matches the marker from the previous request. Any new results are added to a slice of file systems which are looped through later to determine which to delete.

paladin-devops commented on internal/runnerinstall/ecs.go:
```diff
@@ -330,14 +345,76 @@ func (i *ECSRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) e
+	fileSystemsResp, err := efsSvc.DescribeFileSystems(&efs.DescribeFileSystemsInput{
+		CreationToken: nil,
+		FileSystemId:  nil,
+		Marker:        nil,
```
Fixed here! https://github.com/hashicorp/waypoint/pull/3944/commits/0f91f3e67be1ea8a70cf9c95cd4f087635d85f0e

paladin-devops commented on internal/runnerinstall/ecs.go:
```diff
@@ -330,14 +345,76 @@ func (i *ECSRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) e
+					return err
+				}
+				for _, accessPoint := range describeAccessPointsResp.AccessPoints {
+					_, err = efsSvc.DeleteAccessPoint(&efs.DeleteAccessPointInput{AccessPointId: accessPoint.AccessPointId})
```
As I understand it, it is required to delete the access point prior to the mount points.

The purpose for deleting the file system for the runner and not the server is that the server file system contains the server database. If we delete that via the CLI command, then having a chance to recover the database becomes much more difficult. The file system for the runner contains only a token and the runner ID, so it can be deleted with a low risk to impacting the user as they more likely won't need to "recover" a token or runner ID.

catsby replied:
I'm not sure I follow; when we uninstall the server, we do in fact delete the file system, correct? I'm just pointing out the differences here and the server uninstall path. If this code path works correctly and everything is cleaned up then I suppose there's no issue.

paladin-devops commented on internal/runnerinstall/ecs.go:
```diff
@@ -330,14 +345,76 @@ func (i *ECSRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) e
+	for _, fileSystem := range fileSystemsResp.FileSystems {
+		// Check if tags match ID, if so then delete things
+		for _, tag := range fileSystem.Tags {
+			if *tag.Key == "runner-id" && *tag.Value == opts.Id {
```
> use the marker and handle pagination
Updated this to use `Marker` before reading this comment based on your earlier comment.

I'm okay with refactoring the deletion out of the loop, and saving the ID once we have it to be used then. Will update shortly.

paladin-devops commented on internal/runnerinstall/ecs.go:
```diff
@@ -330,14 +345,76 @@ func (i *ECSRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) e
+						return errors.New("after 5 minutes, the file system could" +
+							"not be deleted, because the mount targets weren't deleted")
+					default:
+						_, err = efsSvc.DeleteFileSystem(&efs.DeleteFileSystemInput{FileSystemId: fileSystem.FileSystemId})
```
I agree, will add polling or a sleep here!

paladin-devops committed:
sleep: Update ecs runner install to wait 5 seconds between file system deletion attempts.

paladin-devops commented on internal/runnerinstall/ecs.go:
```diff
@@ -160,12 +158,26 @@ func (i *ECSRunnerInstaller) Install(ctx context.Context, opts *InstallOpts) err
+					Value: aws.String(defaultRunnerTagValue),
+				},
+				{
+					Key:   aws.String("runner-id"),
```
> why is DefaultRunnerTagName exported but defaultRunnerTagValue is not?
Perhaps at one point it may have been used in the `installutil` or `serverinstall` packages, but it appears that if that was the case, it no longer is. Will change to unxported.

Also, I've updated "runner-id" to be a const called "defaultRunnerIdTagName".

paladin-devops committed:
const: Introduce new const for ECS runner uninstall.

Also updates an existing exported const DefaultRunnerTagName to be unexported, since no other package was using it.

briancain commented on internal/runnerinstall/ecs.go:
```diff
@@ -160,12 +160,26 @@ func (i *ECSRunnerInstaller) Install(ctx context.Context, opts *InstallOpts) err
 			}
 
-			if netInfo, err = awsinstallutil.SetupNetworking(ctx, ui, sess, i.Config.Subnets); err != nil {
+			// TODO: Add a port here if the user specified the -liveness-tcp-addr flag
```
Does this need to be turned into a GH issue to be tracked?

paladin-devops replied:
Sure, I'll add one!

paladin-devops replied:
https://github.com/hashicorp/waypoint/issues/4254

briancain commented on internal/runnerinstall/ecs.go:
```diff
@@ -330,15 +347,91 @@ func (i *ECSRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) e
+		}
+	}
+
+DeleteFileSystem:
```
Delete this label? It doesn't look like it's used

paladin-devops replied:
It is used [here](https://github.com/hashicorp/waypoint/pull/3944/files#diff-af12170bdcf18b294209fa40b9af88dc690c261e30f331fc7f520b1798b5a385R385)!

xiaolin-ninja commented on internal/installutil/aws/ecs.go:
```diff
@@ -298,6 +288,7 @@ func CreateService(serviceInput *ecs.CreateServiceInput, ecsSvc *ecs.ECS) (*ecs.
 }
 
 // TODO: Add runner ID as tag
+// SetupExecutionRole
```
Is this comment half-finished with the explanation of the method?

paladin-devops committed:
doc: Add method explanation to SetupExecutionRole.
